### ISSUE TITLE
bug: dist-custom-elements re-assignment causes treeshaking to fail with webpack

### ISSUE THREAD

**Issue by liamdebeasi** [ionic-team/stencil#3191]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/.github/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

2.12.0

### Current Behavior

When using a `dist-custom-elements` Stencil library consumed through a single entry point, Webpack does not treeshake components properly causing all components to be bundled with an application.

This does not happen when consuming components by importing them directly from their built files.

### Expected Behavior

I expect Webpack to only include the components I am using in my application when building.

### Steps to Reproduce

1. Clone repo (make sure you are on the `stencil` branch).
2. Run `cd component-library && npm install && npm run build` to build the Stencil component library.
3. In the `component-library` directory, run `npm pack`.
4. In the root directory, run `npm install`.
5. Open `src/index.js`. Observe that only `AppBadge` is imported.
6. Run `npx webpack`.
7. Run `grep -o -i -E 'Avatar Component|Badge Component' dist/main.js`. This will search the built output for the `"Avatar Component"` and `"Badge Component"` strings. These strings are found in the `app-avatar.js` and `app-badge.js` files and are used to quickly tell if Webpack is bundling the components when building. Observe that both strings appear, meaning both `AppBadge` and `AppAvatar` are bundled with the app.
8. Update `node_modules/component-library/components/app-badge.js` with the following modifications:

```diff
import { HTMLElement, h, Host, proxyCustomElement } from '@stencil/core/internal/client';

let AppBadge$1 = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
  }
  render() {
    return (h(Host, null, "Badge Component"));
  }
};
-AppBadge$1 = /*@__PURE__*/ proxyCustomElement(AppBadge$1, [1, "app-badge"]);
+const AppBadge = /*@__PURE__*/ proxyCustomElement(AppBadge$1, [1, "app-badge"]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["app-badge"];
  components.forEach(tagName => { switch (tagName) {
    case "app-badge":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, AppBadge$1);
      }
      break;
  } });
}

-const AppBadge = AppBadge$1;
const defineCustomElement = defineCustomElement$1;

export { AppBadge, defineCustomElement };

```
9. Update `node_modules/component-library/components/app-avatar.js` with the following modifications:

```diff
import { HTMLElement, h, Host, proxyCustomElement } from '@stencil/core/internal/client';

let AppAvatar$1 = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
  }
  render() {
    return (h(Host, null, "Avatar Component"));
  }
};
-AppAvatar$1 = /*@__PURE__*/ proxyCustomElement(AppAvatar$1, [1, "app-avatar"]);
+const AppAvatar = /*@__PURE__*/ proxyCustomElement(AppAvatar$1, [1, "app-avatar"]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["app-avatar"];
  components.forEach(tagName => { switch (tagName) {
    case "app-avatar":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, AppAvatar$1);
      }
      break;
  } });
}

-const AppAvatar = AppAvatar$1;
const defineCustomElement = defineCustomElement$1;

export { AppAvatar, defineCustomElement };
```
10. Run `npx webpack`.
11. Run `grep -o -i -E 'Avatar Component|Badge Component' dist/main.js`. Observe that only the ``"Badge Component"`` string appears, meaning only `AppBadge` is bundled.

### Code Reproduction URL

https://github.com/liamdebeasi/webpack-reassignment-demo/tree/stencil

### Additional Information

The re-assignment is something Webpack does not account for when building according to https://github.com/webpack/webpack/discussions/14962#discussioncomment-1801168. I have filed an issue on the Webpack repo here: https://github.com/webpack/webpack/issues/14963.

This issue does not impact non-Webpack bundlers such as Rollup.

This is currently impacting Ionic 6 applications where all Ionic components are being added to an app when consuming them through the `@ionic/react` entry point.

**Comment by pratheeshp007** [comment]
Hi, would like to know if this issues is prioritized in an upcoming release.

**Comment by rwaskiewicz** [comment]
This bug has been resolved in [Stencil v2.14.1](https://github.com/ionic-team/stencil/releases/tag/v2.14.1), which was released today (2022.03.07). I'm going to close this issue and mark it as shipped, thanks again!
